### PR TITLE
Added a `CloudManagerSecret` to Cluster API to enable CloudManagers

### PR DESCRIPTION
**Pull request by Richard87**
Added a `CloudManagerSecret` to Cluster API to enable CloudManagers, and specifically add DigitalOceans Cloud-Manager and CSI

Related PR for Bootstrap scripts: https://github.com/kubicorn/bootstrap/pull/23

### CODE REVIEW

**Comment by Richard87** [comment]
This is not ready, I realised the Secret gets "double" base64 encoded, so I need to tweak it a little bit :)

File: apis/cluster/cluster.go [hunk 1]
```diff
@@ -16,6 +16,7 @@ package cluster
 
 import (
 	"encoding/json"
+	"k8s.io/api/core/v1"
```
Comment (xmudrii) [inline apis/cluster/cluster.go]: This should be moved to the second (below) group.

File: apis/cluster/cluster.go [hunk 2]
```diff
@@ -63,6 +64,7 @@ type Cluster struct {
 	// The pattern here says that from an arbitrary deployment, you should be able
 	// to bootstrap anything else you could need. We default to the kubicorn controller.
 	ControllerDeployment *appsv1beta2.Deployment `json:"controllerDeployment,omitempty"`
+	CloudManagerSecret   *v1.Secret              `json:"cloudManagerSecret,omitempty"`
```
Comment (xmudrii) [inline apis/cluster/cluster.go]: We could remove this field to something like `APITokenSecret`, so it's easier to understand what's stored here.

File: cmd/apply.go [hunk 3]
```diff
@@ -208,6 +208,24 @@ func runApply(options *cli.ApplyOptions) error {
...
+		//
+		// Here is where we hook in for the new controller logic
+		// This is exclusive to profiles that have a cloud-manager secret defined
+		//
+		logger.Info("Deploying cloud manager secret: %s", newCluster.CloudManagerSecret.ObjectMeta.Name)
+		err = resourcedeploy.DeployCloudManagerSecret(newCluster)
+		if err != nil {
+			return fmt.Errorf("Unable to deploy cloud manager secret: %v", err)
+		}
+		crdStateStore := crd.NewCRDStore(&crd.CRDStoreOptions{
```
Comment (xmudrii) [inline cmd/apply.go]: This is not needed, along with `crdStateStore.Commit(newCluster)`. This is used by the `cluster-api` and is not needed in this case.

File: profiles/digitalocean/centos_7.go [hunk 4]
```diff
@@ -142,5 +145,15 @@ func NewCentosCluster(name string) *cluster.Cluster {
 	c := cluster.NewCluster(name)
 	c.SetProviderConfig(controlPlaneProviderConfig)
 	c.NewMachineSetsFromProviderConfigs(machineSetsProviderConfigs)
+
+	secret := &apiv1.Secret{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "digitalocean",
+			Namespace: "kube-system",
+		},
+		Data: map[string][]byte{"access-token": []byte(os.Getenv("DIGITALOCEAN_ACCESS_TOKEN"))},
```
Comment (xmudrii) [inline profiles/digitalocean/centos_7.go]: I think you can solve the double-base64 encoding by using `StringData` instead of `Data` here.

**Comment by Richard87** [comment]
Fixed the comments, ad tested a little more

`do` , `digitalocean` and  `do-ubuntu` works nicely :)
`do-centos` failed without any events in `/var/log/cloud-init-output.log` (the file was missing).
`cdo` or `controller-digitalocean-ubuntu` is not tested


It's tested by running this yaml:
```yaml
apiVersion: v1
kind: PersistentVolumeClaim
metadata:
  name: csi-pvc
spec:
  accessModes:
  - ReadWriteOnce
  resources:
    requests:
      storage: 5Gi
  storageClassName: do-block-storage
---
kind: Pod
apiVersion: v1
metadata:
  name: my-csi-app
spec:
  containers:
  - name: my-frontend
    image: busybox
    volumeMounts:
    - mountPath: "/data"
      name: my-do-volume
    command: [ "sleep", "1000000" ]
  volumes:
  - name: my-do-volume
    persistentVolumeClaim:
      claimName: csi-pvc
```
The purpose is to check the CSI and volumes work, and that CSI-DO can read the secret.

**Comment by Richard87** [comment]
New smoketest to also test that DigitalOcean Cloud Manager creates the required LoadBalancer's (works nicely!)

```
apiVersion: v1
kind: PersistentVolumeClaim
metadata:
  name: csi-pvc
spec:
  accessModes:
  - ReadWriteOnce
  resources:
    requests:
      storage: 5Gi
  storageClassName: do-block-storage
---
kind: Pod
apiVersion: v1
metadata:
  name: my-csi-app
spec:
  containers:
  - name: my-frontend
    image: busybox
    volumeMounts:
    - mountPath: "/data"
      name: my-do-volume
    command: [ "sleep", "1000000" ]
  volumes:
  - name: my-do-volume
    persistentVolumeClaim:
      claimName: csi-pvc
---
apiVersion: v1
kind: Service
metadata:
  name: httpbin
  labels:
    app: httpbin
spec:
  type: LoadBalancer
  ports:
  - name: http
    port: 8000
  selector:
    app: httpbin
---
apiVersion: apps/v1
kind: Deployment
metadata:
  name: httpbin
spec:
  replicas: 3
  selector:
    matchLabels:
      app: httpbin
  template:
    metadata:
      labels:
        app: httpbin
        version: v1
    spec:
      containers:
      - image: docker.io/citizenstig/httpbin
        imagePullPolicy: IfNotPresent
        name: httpbin
        ports:
        - containerPort: 8000
```

**Comment by Richard87** [comment]
The loadbalancer and deployments works good. 

*BUT* the Firewall that is created blocks all traffic, but that might be a problem on DO's end

**Pull request merged**